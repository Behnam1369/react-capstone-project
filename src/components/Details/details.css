/* --card-dark-blue: #3d63a4;
--card-light-blue: #4268af; */

.data-details li {
  list-style: none;
}

.data-details li:nth-child(odd) {
  background-color: var(--card-dark-blue);
}

.data-details li:nth-child(even) {
  background-color: var(--card-light-blue);
}
