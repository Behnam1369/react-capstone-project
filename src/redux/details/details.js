import fetchDetailsData from '../../services/fetchDetailsData';

// Actions
const GET_DETAILS = 'GET_DETAILS';
// Reducer
export default function reducer(state = {}, action) {
  switch (action.type) {
    case GET_DETAILS:
      return action.payload;
    default: return state;
  }
}

// Action Creators
export const fetchDetails = (moreInfoLink, name) => async (dispatch) => {
  const payload = await fetchDetailsData(moreInfoLink, name);

  dispatch({
    type: GET_DETAILS,
    payload,
  });
};
