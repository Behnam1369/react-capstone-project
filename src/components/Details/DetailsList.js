import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import { fetchDetails } from '../../redux/details/details';

const DetailsList = ({ name }) => {
  const dispatch = useDispatch();

  const list = [
    {
      id: 0,
      title: 'New Confirmed Cases',
      value: 'newConfirmed',
    },
    {
      id: 1,
      title: 'New Deaths',
      value: 'newDeaths',
    },
    {
      id: 2,
      title: 'Total Confirmed Cases',
      value: 'totalConfirmed',
    },
    {
      id: 3,
      title: 'Total Deaths',
      value: 'totalDeaths',
    },
    {
      id: 4,
      title: 'Total Recovered',
      value: 'totalRecovered',
    },
  ];

  const selectedCountry = useSelector((state) => (
    state.countries.find((country) => country.name === name)
  ));

  useEffect(() => {
    if (selectedCountry !== undefined) {
      dispatch(fetchDetails(selectedCountry.moreInfoLink, selectedCountry.name));
    }
  }, [selectedCountry]);

  const details = useSelector((state) => state.details);
  const navigate = useNavigate();
  return (
    <>
      <button type="button" className="previous-page" onClick={() => navigate(-1)}>Previous Page</button>
      {Object.keys(details).length !== 0 && (
        <ul className="data-details">
          {list.map((item) => (
            <li key={item.id}>
              <p>
                {item.title}
                :
                {' '}
                {details[item.value]}
              </p>
            </li>
          ))}
        </ul>
      )}
    </>
  );
};

DetailsList.propTypes = {
  name: PropTypes.string.isRequired,
};

export default DetailsList;
