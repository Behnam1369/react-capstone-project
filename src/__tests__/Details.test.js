import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import DetailsList from '../components/Details/DetailsList';
import store from '../redux/configureStore';
import { GET_DETAILS } from '../redux/details/details';

const covidData = {
  name: 'Brazil',
  totalConfirmed: 300,
  totalDeaths: 500,
  totalRecovered: 1000,
  newConfirmed: 10,
  newDeaths: 2,
  source: 'Narrativa',
};

const Jsx = () => (
  <BrowserRouter>
    <Provider store={store}>
      <DetailsList name={covidData.name} />
    </Provider>
  </BrowserRouter>
);

const renderCreate = () => renderer.create(Jsx()).toJSON();

const mockCountries = () => {
  render(Jsx());
};

describe('Test Details page', () => {
  store.dispatch({ type: GET_DETAILS, payload: covidData });

  it('Check if the page loads', () => {
    mockCountries();
    const tree = renderCreate();
    expect(screen.getByText(/Total Confirmed/i)).toBeInTheDocument();
    expect(screen.getByText(/Total Deaths/i)).toBeInTheDocument();
    expect(screen.getByText(/Total Recovered/i)).toBeInTheDocument();
    expect(screen.getByText(/New Confirmed Cases/i)).toBeInTheDocument();
    expect(screen.getByText(/New Deaths/i)).toBeInTheDocument();
    expect(tree).toMatchSnapshot();
  });
});
