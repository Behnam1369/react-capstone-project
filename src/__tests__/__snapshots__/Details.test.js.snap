// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Details page Check if the page loads 1`] = `
<ul
  className="data-details p-0 m-0"
>
  <li
    className="card rounded-0 d-flex flex-row justify-content-between p-4 text-white"
  >
    <p
      className="m-0 text-white"
    >
      New Confirmed Cases
    </p>
    <div
      className="d-flex gap-1 align-center"
    >
      10
      <div
        className="d-grid align-content-center"
      >
        <svg
          className="bi bi-arrow-right-circle"
          fill="currentColor"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
  </li>
  <li
    className="card rounded-0 d-flex flex-row justify-content-between p-4 text-white"
  >
    <p
      className="m-0 text-white"
    >
      New Deaths
    </p>
    <div
      className="d-flex gap-1 align-center"
    >
      2
      <div
        className="d-grid align-content-center"
      >
        <svg
          className="bi bi-arrow-right-circle"
          fill="currentColor"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
  </li>
  <li
    className="card rounded-0 d-flex flex-row justify-content-between p-4 text-white"
  >
    <p
      className="m-0 text-white"
    >
      Total Confirmed Cases
    </p>
    <div
      className="d-flex gap-1 align-center"
    >
      300
      <div
        className="d-grid align-content-center"
      >
        <svg
          className="bi bi-arrow-right-circle"
          fill="currentColor"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
  </li>
  <li
    className="card rounded-0 d-flex flex-row justify-content-between p-4 text-white"
  >
    <p
      className="m-0 text-white"
    >
      Total Deaths
    </p>
    <div
      className="d-flex gap-1 align-center"
    >
      500
      <div
        className="d-grid align-content-center"
      >
        <svg
          className="bi bi-arrow-right-circle"
          fill="currentColor"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
  </li>
  <li
    className="card rounded-0 d-flex flex-row justify-content-between p-4 text-white"
  >
    <p
      className="m-0 text-white"
    >
      Total Recovered
    </p>
    <div
      className="d-flex gap-1 align-center"
    >
      1000
      <div
        className="d-grid align-content-center"
      >
        <svg
          className="bi bi-arrow-right-circle"
          fill="currentColor"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
  </li>
</ul>
`;
